[CL-729] Do not show navbar item if corresponding feature is disabled

alexander-cit opened the pull request:
## Checklist

- [x] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [x] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

alexander-cit committed:
[CL-729] Do not show navbar item if corresponding feature is disabled

brentguf commented on back/spec/acceptance/nav_bar_items_spec.rb:
```diff
@@ -37,12 +47,26 @@
+        example_request 'Does not list removed default but feature-disabled NavBarItems' do
+          expect(codes).to include 'home'
+          expect(codes).not_to include 'proposals'
+          expect(codes).not_to include 'events'
```
Where did you disable the `events_widget` feature? Or why do you test on "not including events" in the navbar?

alexander-cit replied:
Currently, `events_widget` feature doesn't control `events` NavBarItem. I just copied this expect from the previous example.

jinjagit replied:
> Where did you disable the `events_widget` feature? Or why do you test on "not including events" in the navbar?

What is the relevance of the `events_widget` to this PR and its ticket?

alexander-cit replied:
Indeed, I think `events_widget` should not be related to this PR. Also, probably it should not control `events` page in the navbar (the meaning seems to be different).

brentguf replied:
No indeed, my mistake. I was just wondering why we are testing for `events` not to be here.

brentguf commented on back/app/policies/nav_bar_item_policy.rb:
```diff
@@ -1,4 +1,6 @@
 class NavBarItemPolicy < ApplicationPolicy
+  FEATURES_CODES = { 'initiatives' => 'proposals' }.freeze
```
Double plural name, should be FEATURE_CODES instead? :-)

alexander-cit replied:
I did it intentionally to show that the keys of the hash are features, and the values are codes.

alexander-cit replied:
(I mean it's not 'feature codes', but 'features and codes')

jinjagit commented on back/spec/acceptance/nav_bar_items_spec.rb:
```diff
@@ -37,12 +47,26 @@
-        expect(json_response[:data].map { |d| d.dig(:attributes, :code) }).to include 'home'
-        expect(json_response[:data].map { |d| d.dig(:attributes, :code) }).not_to include 'events'
+        expect(json_response_body[:data].size).to be > 0
+        expect(codes).to include 'home'
```
I am finding it hard to understand why 'home' is in this list of 'removed default items'. When was it removed and added to this list? (I can't see this happening in this test, but obviously I am missing something)

alexander-cit replied:
The 'home' page wasn't created. That's why it's "removed".

jinjagit replied:
OK, I see. I assumed it would be created by 'default' and would then need to be removed to show in the list of 'removed default items'

alexander-cit replied:
I believe, it was created on production in a migration script, when we introduced this feature.

jinjagit reviewed (approved):
LGTM, though I have asked a question about a test that I didn't really understand completely.

alexander-cit committed:
[CL-729] Disable addButton if itemsLength is greater than limit

Consider this scenario (there can be more):
1. There are 7 pages in navbar including proposals
2. Proposals are disabled. There are 6 pages now
3. Admin adds a page to navbar. There are 7 pages again
4. Proposals are enabled. There are 8 pages now. Admin can add more pages.

Practically, there's no limit anymore. We should prevent it.

alexander-cit commented:
@brentguf I tested 7 pages in navbar and handled it in this commit https://github.com/CitizenLabDotCo/citizenlab/pull/1964/commits/86cf0301aad4ab4161d97e5f83a4bdd4f2b60f18

brentguf commented:
> @brentguf I tested 7 pages in navbar and handled it in this commit [86cf030](https://github.com/CitizenLabDotCo/citizenlab/commit/86cf0301aad4ab4161d97e5f83a4bdd4f2b60f18)

We actually need to make sure adding the new nav bar item is not possible if we already have 7 items. In the BE and/or FE. In the FE by probably not allowed enabling of the proposals feature if we have 7 navbar items (this is a different ticket though). Or in the BE by not adding it to the visible items, but in the non-visible items list.

If we already have 7 items and I enable proposals, will it be added automatically by the back-end and display 8 items?

alexander-cit commented:
@brentguf 

> not allowed enabling of the proposals feature if we have 7 navbar items

For me, it looks too hard from UX point of view.

> not adding it to the visible items, but in the non-visible items list

Currently, there's no such thing as visible/non-visible items (except this trick with disabling the `proposals` feature). We can create or delete navBarItem. If we delete it disabling the feature, it means we lose custom translations for `title` permanently. Which is probably not great, but not that terrible.

> it will be added automatically by the back-end and display 8 items. No?

Yes, there will be 8 items. Do you think it's a big deal? Anyway, this value is not precisely calculated, it's almost random. NavBarItems look completely different for different languages/copies and we don't control it.

alexander-cit commented:
@brentguf Actually, I don't mind your proposed solutions. Just wanted to point out they have some downsides apart from additional work. So, if you still think we should somehow change it, I think we could do it.

alexander-cit commented:
E.g. to delete a page disabling the feature, we need to introduce really tricky connections between different modules of BE code. I would rather avoid it in this case.

brentguf commented:
Let me think about it some more, but next week.

brentguf commented:
@alexander-cit: I'm syncing with Ben next week on what would be best UX-wise. This PR is not urgent, so that's OK I guess.

brentguf commented:
@alexander-cit I think you can merge this as the next part of the solution will normally be in the front-end.

So to summarize: if I turn on proposals, its corresponding nav bar item will be in the list of nav bar items. If I turn the feature off, it will not be in this list (but it won't be deleted). If that's correct, you can merge!

alexander-cit commented:
> If I turn the feature off, it will not be in this list (but it won't be deleted)

Yes, you're absolutely right!

alexander-cit committed:
Merge branch 'master' into CL-729-remove-nav-bar-item-if-feature-disabled
